syntax = "proto3";

package training;

service TrainingService {
  rpc RegisterWorker(StartRequest) returns (StartResponse);
  rpc GetInitialWeights(WeightRequest) returns (WeightResponse);
  rpc PushGradients(GradientData) returns (Ack);
  rpc GetUpdatedWeights(WeightRequest) returns (WeightResponse);
  rpc ReportMetrics(MetricData) returns (Ack);
  rpc FinishTraining(FinishRequest) returns (Ack);
}

message StartRequest {
  string worker_id = 1;
}
message StartResponse {
  bool ready = 1;
  int32 start = 2;
  int32 end = 3;
  int32 report_step = 4;
}

message Ack {
  bool success = 1;
  string message = 2;
}

message MetricData {
  string worker_id = 1;
  int32 epoch = 2;
  int32 step = 3;
  float loss = 4;
  float accuracy = 5;
  string timestamp = 6;
}
message FinishRequest {
  string worker_id = 1;
}

message WeightRequest {
  string worker_id = 1;
  int32 epoch = 2; // Opcional, para tracking
}
message WeightResponse {
  string weights_path = 1;
}

message GradientData {
  string worker_id = 1;
  int32 step = 2;
  bytes serialized_gradients = 3; // Gradientes serializados
}